parser.go

```diff
@@ -205,251 +202,99 @@ func (p *parser) parseIP6(t *token) (bool, Result, error) {
-			hits <- hit
-			wg.Done()
-		}(mx.Mx, ip4Mask, ip6Mask)
-	}
-
-	go func() {
-		wg.Wait()
-		close(hits)
-	}()
-
-	for hit := range hits {
-		if hit {
-			return true, result, nil
+		switch len(ip) {
+		case net.IPv4len:
+			n.Mask = ip4Mask
+		case net.IPv6len:
+			n.Mask = ip6Mask
 		}
-	}
-	return false, result, nil
+		return n.Contains(p.IP)
+	})
+	return found, result, err
 }
 
 func (p *parser) parseInclude(t *token) (bool, Result, error) {
-	result, _ := matchingResult(t.qualifier)
 	domain := t.value
-	if isEmpty(&domain) {
+	if domain == "" {
 		return true, Permerror, SyntaxError{t, errors.New("empty domain")}
 	}
-	matchesInclude := false
-	includeResult, _, err := CheckHost(p.IP, domain, p.Sender, p.Config)
-	if err != nil {
-		return false, None, SyntaxError{t, err}
+	theirResult, _, err := checkHost(p.IP, domain, p.Sender, p.resolver)
+	if err != nil && err != errDNSLimitExceeded {
+		return true, None, SyntaxError{t, err}
 	}
 
 	// it's all fine
-	switch includeResult {
+	switch theirResult {
 	case Pass:
-		matchesInclude = true
+		ourResult, _ := matchingResult(t.qualifier)
+		return true, ourResult, nil
 	case Fail, Softfail, Neutral:
-		matchesInclude = false
+		return false, None, nil
 	case Temperror:
-		matchesInclude = false
-		result = Temperror
-	case Permerror, None:
-		matchesInclude = false
-		result = Permerror
-	}
-
-	if matchesInclude {
-		return true, result, nil
+		return true, Temperror, nil
+	case Permerror:
+		return true, Permerror, nil
+	case None:
+		fallthrough
+	default:
+		return false, None, nil
 	}
-	return false, None, nil
 }
 
 func (p *parser) parseExists(t *token) (bool, Result, error) {
-	result, _ := matchingResult(t.qualifier)
 	resolvedDomain, err := parseMacroToken(p, t)
 	if err != nil {
 		return true, Permerror, SyntaxError{t, err}
-	} else if isEmpty(&resolvedDomain) {
+	}
+	if resolvedDomain == "" {
 		return true, Permerror, SyntaxError{t, errors.New("empty domain")}
 	}
-	resolvedDomain = normalizeHost(resolvedDomain)
-	var queries [2]dns.Msg
 
-	queries[0].SetQuestion(resolvedDomain, dns.TypeA)
-	queries[1].SetQuestion(resolvedDomain, dns.TypeAAAA)
-	for _, query := range queries {
-		c := new(dns.Client)
-		response, _, err := c.Exchange(&query, p.Config.dnsAddr)
-		if err != nil {
-			return true, Temperror, SyntaxError{t, err}
-		}
+	result, _ := matchingResult(t.qualifier)
 
-		if response != nil && response.Rcode != dns.RcodeSuccess {
-			if response.Rcode == dns.RcodeNameError {
-				return false, result, nil
-			}
-			return true, Temperror, SyntaxError{t,
-				fmt.Errorf("unsuccessful DNS response, code %d", response.Rcode)}
-		}
-		/* We can check prematurely and avoid further DNS calls if matching
-		 * hosts were already found.
-		 */
-		if len(response.Answer) > 0 {
-			return true, result, nil
-		}
+	found, err := p.resolver.Exists(normalizeFQDN(resolvedDomain))
+	switch err {
+	case nil:
+		return found, result, nil
+	case ErrDNSPermerror:
+		return false, result, nil
```

Comment:
return false, as we want recursive error to bubble up. Otherwise, this term will be discarded and  Parser.parse() will proceed to another term (section 4.6)